### PR TITLE
JIT: Update gtHasLocalsWithAddrOp to use GenTreeVisitor

### PR DESCRIPTION
**Pull request by jakobbotsch**
Also change it to check for any local, not just GT_LCL_VAR.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib, simple change with [no diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=18123&view=ms.vss-build-web.run-extensions-tab) and some small TP improvements.

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -2724,48 +2724,48 @@ bool GenTree::Compare(GenTree* op1, GenTree* op2, bool swapOK)
...
+        }
 
-    desc.comp            = this;
-    desc.hasAddrTakenLcl = false;
+        fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
+        {
+            LclVarDsc* varDsc = m_compiler->lvaGetDesc((*use)->AsLclVarCommon());
+            if (varDsc->lvHasLdAddrOp || varDsc->IsAddressExposed())
+            {
+                HasAddrTakenLocal = true;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/gentree.cpp:2759]: Shouldn't we still early out like before?
```suggestion
                HasAddrTakenLocal = true;
                return WALK_ABORT;
```

Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:2759]: Whoops, yes, good catch.

**Comment by jakobbotsch** [comment]
@AndyAyersMS Can you take another look?

**Comment by jakobbotsch** [comment]
Diffs look spurious (#53773). I think the #53773 problem is more pronounced now that we have frozen objects too, we probably need a few JIT-EE interface changes to make sure that we can use the captured field addresses/object addresses during SPMI (JIT probably needs to ask the EE about values instead of dereferencing them directly, should probably help R2R/NAOT too). cc @EgorBo

**Comment by jakobbotsch** [comment]
Ping @AndyAyersMS

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**